### PR TITLE
Add transformer doc

### PR DESCRIPTION
**Pull request by ehoelzl**

### CODE REVIEW

File: benchmark-tasks.rst [hunk 1]
```diff
@@ -187,18 +193,116 @@ TODO
...
+    This avoid having `nan/inf` in weights and gradients, due to overflows and underflows.
+
+#. **Selection of Framework & Systems**
+    We currently only have this reference implementation in PyTorch. For the systems, kubernetes allows easy transferability of our code.
+    While initial results reported are from google kubernetes engine, AWS will be supported very soon.
+
+#. **Environments for Scaling Task**
+    For the scaling task, we use `n1-standard-4 <https://cloud.google.com/compute/pricing>`_ type instances with 50GB disk size.
+    There is only one worker per node; each worker uses 2.5 cpus. The bandwidth between two nodes is around 7.5Gbit/s.
+    Openmpi, NCCL or GLOO are used for communication. No accelerators are used for this task.
```
Comment (martinjaggi) [inline benchmark-tasks.rst]: MPI

File: benchmark-tasks.rst [hunk 2]
```diff
@@ -187,18 +193,116 @@ TODO
...
+
+#. **Data Preprocessing**
+    The data needs to be downloaded and pre-processed and tokenized using the pre-processing script
+    `mlbench_core/dataset/nlp/pytorch/wmt17/preprocess/preprocess.py` before training.
+    The pre-processed data is available on our `S3 storage <https://storage.googleapis.com/mlbench-datasets/translation/wmt17_en_de.tar.gz>`_
+
+#. **Mixed Precision Training**
+    In order to have faster backward and forward passes, our model's weights and gradients are cast into Float16 prior to training.
+    Float32 weights are still kept in memory and used by the optimizer to update weights. We use our own `FP16Optimizer`.
+    Since Float16 has lower precision than Float32, it is necessary to have a loss scaling, that scales the loss before each backward pass.
```
Comment (martinjaggi) [inline benchmark-tasks.rst]: add a reference/link for more details of how it's scaled?

File: benchmark-tasks.rst [hunk 3]
```diff
@@ -187,18 +193,116 @@ TODO
...
+    - maximum epochs: 10
+    - learning rate
+
+      + initial learning rate :math:`\eta` : 0.0
+      + base learning rate :math:`\eta`: 1.976e-3
+      + decay: We decay by :math:`\sqrt{N}` after warmup
+      + scaling and warmup: We use 1000 warmup steps, where the learning rate is linearly increased from
+        `initial_learning_rate` to `base_learning_rate`
+
+    - betas: (0.9, 0.98)
```
Comment (martinjaggi) [inline benchmark-tasks.rst]: for adam. also eps?

File: benchmark-tasks.rst [hunk 4]
```diff
@@ -187,18 +193,116 @@ TODO
...
+    torch's native `MultiheadAttention`
+
+#. **Dataset**
+    The `WMT-17 <http://www.statmt.org/wmt17/>`_
+    dataset containing a set translated sentences from multiple languages.
+    We exclusively use English-German translation from this dataset.
+
+
+#. **Training Algorithm**
+    We use standard Adam as the optimizer (that is distributed Adam with synchronous all-reduce communication before each update step).
```
Comment (martinjaggi) [inline benchmark-tasks.rst]: mention if the adam statistics (second moment statistics) are averaged in each step as well (i assume so)

File: benchmark-tasks.rst [hunk 5]
```diff
@@ -187,18 +193,116 @@ TODO
...
 
 
+3 Natural Language Processing
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+3a. Language Modeling
+#####################
+(TODO)
+
+3b. Translation EN-DE (GNMT, WMT14)
```
Comment (martinjaggi) [inline benchmark-tasks.rst]: say LSTM here, and GNMT later when giving specific details on the architecture.
@Panaetius can you add details here later?

File: benchmark-tasks.rst [hunk 6]
```diff
@@ -187,18 +193,116 @@ TODO
...
+    `Attention Is All You need <https://arxiv.org/abs/1706.03762>`_ that. The implementation is based on
+    a combination of NVIDIA's implementation of `fairseq <https://github.com/pytorch/fairseq>`_ 's transformer.
+    Our implementation differs from MLPerf's in one subtle way: the `FusedLayerNorm` layers are changed to native
+    torch `LayerNorm`, as its performance has increased since. Also, instead of using `FusedAdam`, we use `Adam`.
+    One part of the `MultiheadAttention` module needs a cuda extension, that makes training significantly faster than
+    torch's native `MultiheadAttention`
+
+#. **Dataset**
+    The `WMT-17 <http://www.statmt.org/wmt17/>`_
+    dataset containing a set translated sentences from multiple languages.
```
Comment (martinjaggi) [inline benchmark-tasks.rst]: typo set of

**Review by martinjaggi — CHANGES_REQUESTED** [review]
nice work!

File: benchmark-tasks.rst (hunk 5 above)
Comment (ehoelzl) [reply benchmark-tasks.rst]: I can add the details, as I have done the code for this part too. Will add it later this week (in a new PR) once everything is wrapped up and merged.

File: benchmark-tasks.rst (hunk 4 above)
Comment (ehoelzl) [reply benchmark-tasks.rst]: I don't think they are. We simply average the gradients, but each worker has his own set of param groups. Now since all workers perform the same weight updates (with the same gradients) and are all initialized in the same way, the state of each worker's optimizer should be the same. (Correct me if I'm wrong)

Comment (martinjaggi) [reply benchmark-tasks.rst]: you're right, all good (and no communication overhead this way which is a main advantage)
